admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -0,0 +1,196 @@
+};
+use prost_ext::MessageExt;
+use rand::seq::SliceRandom;
+use std::collections::HashMap;
+
+/// Struct that owns all dependencies required by upstream proxy routine
+pub(crate) struct UpstreamProxy {
+    ac_config: AdmissionControlConfig,
+    network_sender: AdmissionControlNetworkSender,
+    peer_info: HashMap<PeerId, bool>,
+    /// used to process client requests
+    client_events: mpsc::UnboundedReceiver<(
+        SubmitTransactionRequest,
+        oneshot::Sender<failure::Result<SubmitTransactionResponse>>,
+    )>,
+    role: RoleType,
+    /// AC Client
+    pub client: AdmissionControlClient,
+}
+
+impl UpstreamProxy {
+    /// bootstrap of UpstreamProxy
+    pub fn new(
+        config: &NodeConfig,
+        network_sender: AdmissionControlNetworkSender,
+        client_events: mpsc::UnboundedReceiver<(
+            SubmitTransactionRequest,
+            oneshot::Sender<failure::Result<SubmitTransactionResponse>>,
+        )>,
+        client: AdmissionControlClient,
+    ) -> Self {
+        let upstream_peer_ids = config.get_upstream_peer_ids();
+        let peer_info: HashMap<_, _> = upstream_peer_ids
+            .iter()
+            .map(|peer_id| (*peer_id, true))
+            .collect();
+
+        Self {
+            ac_config: config.admission_control.clone(),
+            network_sender,
+            peer_info,
+            client_events,
+            role: config.get_role(),
+            client,
+        }
+    }
+
+    /// main routine. starts sync coordinator that listens for CoordinatorMsg
+    pub async fn process_network_messages(
+        mut self,
+        network_events: Vec<AdmissionControlNetworkEvents>,
+    ) {
+        let mut events = select_all(network_events).fuse();
+
+        loop {
+            ::futures::select! {
+                (mut msg, mut callback) = self.client_events.select_next_some() => {
+                    let result = self.submit_transaction_upstream(msg).await;
+                    let _ = callback.send(result);
+                },
+                network_event = events.select_next_some() => {
+                    match network_event {
+                        Ok(event) => {
+                            match event {
+                                Event::NewPeer(peer_id) => {
+                                    debug!("[admission control] new peer {}", peer_id);
+                                    self.new_peer(peer_id);
+                                }
+                                Event::LostPeer(peer_id) => {
+                                    debug!("[admission control] lost peer {}", peer_id);
+                                    self.lost_peer(peer_id);
+                                }
+                                Event::RpcRequest((peer_id, mut message, callback)) => {
+                                    if let Some(AdmissionControlMsg_oneof::SubmitTransactionRequest(request)) = message.message {
```

Comment:
do you need to use oneof syntax? I thought there's some `.has_` syntax available

Reply:
I was following how Consensus does it with ConsensusMsg_oneof